cmd/tempo-vulture/main.go

```diff
@@ -45,95 +43,189 @@ func init() {
+			// don't query traces before retention
+			if (currentTime - startTime) > int64(tempoRetentionDuration/time.Second) {
+				startTime = currentTime - int64(tempoRetentionDuration/time.Second)
 			}
+
+			// pick past slot and re-generate trace
+			rand.Seed((generateRandomInt(startTime, currentTime) / slot) * slot)
+			hexID := fmt.Sprintf("%016x%016x", rand.Int63(), rand.Int63())
+
+			// query the trace
+			metrics, err := queryTempoAndAnalyze(tempoQueryURL, hexID)
+			if err != nil {
+				glog.Error("error querying Tempo ", err)
+				metricErrorTotal.Inc()
+				metricTracesErrors.WithLabelValues("failed").Inc()
+				continue
+			}
+
+			metricTracesInspected.Add(float64(metrics.requested))
+			metricTracesErrors.WithLabelValues("notfound").Add(float64(metrics.notfound))
+			metricTracesErrors.WithLabelValues("missingspans").Add(float64(metrics.missingSpans))
 		}
 	}()
 
 	http.Handle(prometheusPath, promhttp.Handler())
 	log.Fatal(http.ListenAndServe(prometheusListenAddress, nil))
 }
 
-func queryTempoAndAnalyze(baseURL string, backoff time.Duration, traceIDs []string) (*traceMetrics, error) {
-	tm := &traceMetrics{
-		requested: len(traceIDs),
+func newJaegerGRPCClient(endpoint string) (*jaeger_grpc.Reporter, error) {
+	// remove scheme and port
+	u, err := url.Parse(endpoint)
+	if err != nil {
+		return nil, err
 	}
+	host, _, err := net.SplitHostPort(u.Host)
+	if err != nil {
+		return nil, err
+	}
+	// new jaeger grpc exporter
+	conn, err := grpc.Dial(host+":14250", grpc.WithInsecure())
+	if err != nil {
+		return nil, err
+	}
+	logger, err := zap.NewDevelopment()
+	if err != nil {
+		return nil, err
+	}
+	return jaeger_grpc.NewReporter(conn, nil, logger), err
+}
 
-	for _, id := range traceIDs {
-		time.Sleep(backoff)
+func generateRandomString() string {
+	var letters = []rune("abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ")
 
-		glog.Error("tempo url ", baseURL+"/api/traces/"+id)
-		trace, err := util.QueryTrace(baseURL, id, tempoOrgID)
-		if err == util.ErrTraceNotFound {
-			glog.Error("trace not found ", id)
-			tm.notfound++
-			continue
-		}
-		if err != nil {
-			return nil, err
-		}
+	s := make([]rune, generateRandomInt(5, 20))
+	for i := range s {
+		s[i] = letters[rand.Intn(len(letters))]
+	}
+	return string(s)
+}
 
-		if len(trace.Batches) == 0 {
-			glog.Error("trace not found", id)
-			tm.notfound++
-			continue
-		}
+func generateRandomTags() []*thrift.Tag {
+	var tags []*thrift.Tag
+	for i := int64(0); i < generateRandomInt(1, 5); i++ {
+		value := generateRandomString()
+		tags = append(tags, &thrift.Tag{
+			Key:  generateRandomString(),
+			VStr: &value,
+		})
+	}
+	return tags
+}
 
-		// iterate through
-		if hasMissingSpans(trace) {
-			glog.Error("has missing spans", id)
-			tm.missingSpans++
-		}
+func generateRandomLogs() []*thrift.Log {
+	var logs []*thrift.Log
+	for i := int64(0); i < generateRandomInt(1, 5); i++ {
```

Comment:
i think `generateRandomInt()` will be called on every loop evaluation.  this technically works but is kind of weird.  let's move it up a line:

```
count := generateRandomInt(1, 5)
for i := int64(0); i < count; i++ {
...
```

Reply:
Didn't know that! Changed :)